adapters/rest-vertx/src/main/java/org/eclipse/hono/adapter/rest/VertxBasedRestProtocolAdapter.java

```diff
@@ -12,54 +12,94 @@
-    protected void addRoutes(final Router router) {
+    protected final void addRoutes(final Router router) {
+        if (!getConfig().isAuthenticationRequired()) {
+            LOG.warn("authentication of devices switched off");
+        } else {
+            setupBasicAuth(router);
+        }
         addTelemetryApiRoutes(router);
         addEventApiRoutes(router);
     }
 
-    private void addTelemetryApiRoutes(final Router router) {
+    private final void setupBasicAuth(final Router router) {
+        router.route().handler(BasicAuthHandler.create(this));
+    }
 
+    private final void addTelemetryApiRoutes(final Router router) {
         // route for uploading telemetry data
         router.route(HttpMethod.PUT, String.format("/telemetry/:%s/:%s", PARAM_TENANT, PARAM_DEVICE_ID))
-            .handler(ctx -> uploadTelemetryMessage(ctx, getTenantParam(ctx), getDeviceIdParam(ctx)));
+                .handler(ctx -> uploadTelemetryMessage(ctx, getTenantParam(ctx), getDeviceIdParam(ctx)));
     }
 
-    private void addEventApiRoutes(final Router router) {
-
+    private final void addEventApiRoutes(final Router router) {
         // route for sending event messages
         router.route(HttpMethod.PUT, String.format("/event/:%s/:%s", PARAM_TENANT, PARAM_DEVICE_ID))
-            .handler(ctx -> uploadEventMessage(ctx, getTenantParam(ctx), getDeviceIdParam(ctx)));
+                .handler(ctx -> uploadEventMessage(ctx, getTenantParam(ctx), getDeviceIdParam(ctx)));
     }
 
-    private static String getTenantParam(final RoutingContext ctx) {
+    private final static String getTenantParam(final RoutingContext ctx) {
         return ctx.request().getParam(PARAM_TENANT);
     }
 
-    private static String getDeviceIdParam(final RoutingContext ctx) {
+    private final static String getDeviceIdParam(final RoutingContext ctx) {
         return ctx.request().getParam(PARAM_DEVICE_ID);
     }
 
     /**
      * Returns {@code null} to disable status resource.
      */
     @Override
-    protected String getStatusResourcePath() {
+    protected final String getStatusResourcePath() {
         return null;
     }
+
+    @Override
+    public final void authenticate(JsonObject authInfo, Handler<AsyncResult<User>> resultHandler) {
+        String username = authInfo.getString(USERNAME_PROPERTY);
+        String password = authInfo.getString(PASSWORD_PROPERTY);
+        UsernamePasswordCredentials credentials = UsernamePasswordCredentials.create(username,
+                password, getConfig().isSingleTenant());
+        if (credentials == null) {
+            resultHandler.handle(Future.failedFuture("login failed"));
+        } else {
+            this.validateCredentialsForDevice(credentials).setHandler(attempt -> {
+                if (attempt.succeeded()) {
+                    resultHandler.handle(Future.succeededFuture(null));
+                } else {
+                    resultHandler.handle(Future.failedFuture("login failed"));
```

Comment:
Ditto as above

Reply:
Here I could do a bit better then above. I reckon we don't need to pass more information to the handler. We won't show people why the authentication failed anyways only that the given credentials are incorrect. The "BadRequest" check is done by the vertx component